File: 백준/Silver/24479. 알고리즘 수업 － 깊이 우선 탐색 1/알고리즘 수업 － 깊이 우선 탐색 1.py
import sys

sys.setrecursionlimit(150000)

def dfs(start):
  global cnt
  visited[start] = cnt
  graph[start].sort()

  for i in graph[start]:
    if visited[i] == 0:
      cnt += 1
      dfs(i)

n, m, r = map(int, sys.stdin.readline().split())
graph = [[] for _ in range(n + 1)]
visited = [0] * (n + 1)
cnt = 1

for _ in range(m):
  u, v = map(int, sys.stdin.readline().split())
  graph[u].append(v)
  graph[v].append(u)

dfs(r)

for i in range(1, n + 1):
  print(visited[i])

File: 백준/Silver/24479. 알고리즘 수업 － 깊이 우선 탐색 1/알고리즘 수업 － 깊이 우선 탐색 1.cc
#include <iostream>
#include <vector>
#include <algorithm>

std::vector<int> graph[100001];
bool visited[100001];
int result[100001];
int count = 0;

static void dfs(int node)
{
	if (visited[node]) return;

	visited[node] = true;
	result[node] = ++count;

	for (int next : graph[node])
	{
		dfs(next);
	}
}

int main()
{
	std::cin.tie(NULL);
	std::ios_base::sync_with_stdio(false);

	int n, m, r;
	std::cin >> n >> m >> r;

	for (int i = 0; i < m; ++i)
	{
		int u, v;
		std::cin >> u >> v;

		graph[u].push_back(v);
		graph[v].push_back(u);
	}

	for (int i = 1; i <= n; ++i)
	{
		sort(graph[i].begin(), graph[i].end());
	}

	dfs(r);

	for (int i = 1; i <= n; ++i)
	{
		std::cout << result[i] << "\n";
	}

	return 0;
}